File: 03 March/03-29/is_valid_isbn10.py
def is_valid_isbn10(s):
    hyphens_removed = s.replace("-", "")

    if not re.match(r"^\d{9}(\d|X)$", hyphens_removed):
        return False

    total = 0

    for index, digit in enumerate(hyphens_removed):
        total += (10 if digit == "X" else int(digit)) * (index + 1)

    return not total % 11

# --- TEST SUITE ---

tests_text = r'''
1. is_valid_isbn10("0-306-40615-2") should return True.
2. is_valid_isbn10("0-306-40615-1") should return False.
3. is_valid_isbn10("0-8044-2957-X") should return True.
4. is_valid_isbn10("X-306-40615-2") should return False.
5. is_valid_isbn10("0-6822-2589-4") should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-29/isValidIsbn10.js
function isValidIsbn10(str) {
  const hyphensRemoved = str.replaceAll("-", "");

  if (!/^\d{9}(\d|X)$/.test(hyphensRemoved))
    return false;

  return !Boolean(hyphensRemoved
    .split("")
    .reduce((total, digit, index) => total + (digit === "X" ? 10 : digit) * (index + 1), 0)
    % 11);
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isValidIsbn10("0-306-40615-2") should return true.
2. isValidIsbn10("0-306-40615-1") should return false.
3. isValidIsbn10("0-8044-2957-X") should return true.
4. isValidIsbn10("X-306-40615-2") should return false.
5. isValidIsbn10("0-6822-2589-4") should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);